storage/innobase/fil/fil0fil.cc

```diff
@@ -942,111 +934,87 @@ fil_space_free(
-@param name       tablespace name
-@param id         tablespace identifier
-@param flags      tablespace flags
-@param purpose    tablespace purpose
-@param crypt_data encryption information
-@param mode       encryption mode
-@param opened     true if space files are opened
-@return pointer to created tablespace, to be filled in with add()
-@retval nullptr on failure (such as when the same tablespace exists) */
-fil_space_t *fil_space_t::create(ulint id, ulint flags,
-                                 fil_type_t purpose,
-				 fil_space_crypt_t *crypt_data,
-				 fil_encryption_t mode,
-				 bool opened)
-{
-	fil_space_t*	space;
-
-	mysql_mutex_assert_owner(&fil_system.mutex);
-	ut_ad(fil_system.is_initialised());
-	ut_ad(fil_space_t::is_valid_flags(flags & ~FSP_FLAGS_MEM_MASK, id));
-	ut_ad(srv_page_size == UNIV_PAGE_SIZE_ORIG || flags != 0);
-
-	DBUG_EXECUTE_IF("fil_space_create_failure", return(NULL););
-
-	fil_space_t** after = reinterpret_cast<fil_space_t**>(
-		&fil_system.spaces.cell_get(id)->node);
-	for (; *after; after = &(*after)->hash) {
-		ut_a((*after)->id != id);
-	}
-
-	/* FIXME: if calloc() is defined as an inline function that calls
-	memset() or bzero(), then GCC 6 -flifetime-dse can optimize it away */
-	*after = space = new (ut_zalloc_nokey(sizeof(*space))) fil_space_t;
-
-	space->id = id;
-
-	UT_LIST_INIT(space->chain, &fil_node_t::chain);
-
-	space->purpose = purpose;
-	space->flags = flags;
-
-	space->crypt_data = crypt_data;
-	space->n_pending.store(CLOSING, std::memory_order_relaxed);
-
-	DBUG_LOG("tablespace", "Created metadata for " << id);
-	if (crypt_data) {
-		DBUG_LOG("crypt",
-			 "Tablespace " << id
-			 << " encryption " << crypt_data->encryption
-			 << " key id " << crypt_data->key_id
-			 << ":" << fil_crypt_get_mode(crypt_data)
-			 << " " << fil_crypt_get_type(crypt_data));
-	}
-
-	space->latch.SRW_LOCK_INIT(fil_space_latch_key);
-
-	if (opened)
-	  fil_system.add_opened_last_to_space_list(space);
-	else
-          fil_system.space_list.push_back(*space);
-
-	switch (id) {
-	case 0:
-		ut_ad(!fil_system.sys_space);
-		fil_system.sys_space = space;
-		break;
-	case SRV_TMP_SPACE_ID:
-		ut_ad(!fil_system.temp_space);
-		fil_system.temp_space = space;
-		break;
-	default:
-		ut_ad(purpose != FIL_TYPE_TEMPORARY);
-		if (UNIV_LIKELY(id <= fil_system.max_assigned_id)) {
-			break;
-		}
-		if (UNIV_UNLIKELY(srv_operation == SRV_OPERATION_BACKUP)) {
-			break;
-		}
-		if (!fil_system.space_id_reuse_warned) {
-			ib::warn() << "Allocated tablespace ID " << id
-				<< ", old maximum was "
-				<< fil_system.max_assigned_id;
-		}
+fil_space_t::fil_space_t(uint32_t id, ulint flags, bool being_imported,
+                         fil_space_crypt_t *crypt_data) noexcept :
+  id(id), crypt_data(crypt_data), being_imported(being_imported), flags(flags)
+{
+  UT_LIST_INIT(chain, &fil_node_t::chain);
+#ifndef SUX_LOCK_GENERIC
+  memset((void*) &latch, 0, sizeof latch);
+#endif
+  latch.SRW_LOCK_INIT(fil_space_latch_key);
```

Comment:
Is it safer to keep this memset unconditional ? It is fine if you think SUX_LOCK_GENERIC can handle it.

Reply:
The memset is unnecessary for `SUX_LOCK_GENERIC`. Most notably, this includes the following platforms:
 * Apple macOS
 * IBM AIX
 * Sun Solaris and its derivatives (none of which we have on our CI)

Sure, we could make the zero-initialization unconditional. It is conditional because the `futex` based `srw_lock` does not initialize anything in its constructor and unlike many other places, here we actually need some initialization.

At least on macOS and IBM AIX, based on 4f8803c036a23fb550df6f3d7a669bf71f265f57 we know that the following initialization is not a zero initialization.

It could be cleaner to introduce an `inline` member function (say `pre_init()`) for this additional initialization. The implementation for `SUX_LOCK_GENERIC` would be empty.